src/wrap_mempool.cpp

```diff
@@ -275,22 +277,234 @@ namespace
+
+    public:
+      svm_allocator(std::shared_ptr<pyopencl::context> const &ctx,
+          cl_uint alignment, cl_svm_mem_flags flags=CL_MEM_READ_WRITE,
+          pyopencl::command_queue *queue=nullptr)
+        : m_context(ctx), m_alignment(alignment), m_flags(flags)
+      {
+        if (queue)
+          m_queue.set(queue->data());
+      }
+
+      svm_allocator(svm_allocator const &src)
+      : m_context(src.m_context), m_alignment(src.m_alignment),
+      m_flags(src.m_flags)
+      { }
+
+      virtual ~svm_allocator()
+      { }
+
+      virtual bool is_deferred() const
+      {
+        // FIXME: I don't know whether that's true.
+        return false;
+      }
+
+      std::shared_ptr<pyopencl::context> context() const
+      {
+        return m_context;
+      }
+
+      pointer_type allocate(size_type size)
+      {
+        if (size == 0)
+          return { nullptr, nullptr };
+
+        PYOPENCL_PRINT_CALL_TRACE("clSVMalloc");
+        return {
+          clSVMAlloc(m_context->data(), m_flags, size, m_alignment),
+          pyopencl::command_queue_ref(m_queue.data())
+        };
+      }
+
+      virtual pointer_type hand_out_existing_block(pointer_type &&p)
+      {
+        if (m_queue.is_valid())
+        {
+          if (p.queue.is_valid())
+          {
+            // make sure synchronization promises stay valid in new queue
+            cl_event evt;
+
+            PYOPENCL_CALL_GUARDED(clEnqueueMarker, (p.queue.data(), &evt));
+            PYOPENCL_CALL_GUARDED(clEnqueueMarkerWithWaitList,
+                (m_queue.data(), 1, &evt, nullptr));
+          }
+          p.queue.set(m_queue.data());
+        }
+        return std::move(p);
+      }
+
+      void free(pointer_type &&p)
+      {
+        if (p.queue.is_valid())
+        {
+          PYOPENCL_CALL_GUARDED_CLEANUP(clEnqueueSVMFree, (
+                p.queue.data(), 1, &p.ptr,
+                nullptr, nullptr,
+                0, nullptr, nullptr));
+          p.queue.reset();
+        }
+        else
+        {
+          PYOPENCL_PRINT_CALL_TRACE("clSVMFree");
+          clSVMFree(m_context->data(), p.ptr);
+        }
+      }
+
+      void try_release_blocks()
+      {
+        pyopencl::run_python_gc();
+      }
+  };
 
+  // }}}
 
 
+  // {{{ svm_allocator_call
+
+  inline
+  pyopencl::svm_allocation *svm_allocator_call(svm_allocator &alloc, size_t size)
+  {
+    int try_count = 0;
+    while (true)
+    {
+      try
+      {
+        svm_held_pointer mem(alloc.allocate(size));
+        return new pyopencl::svm_allocation(alloc.context(), mem.ptr, size);
+      }
+      catch (pyopencl::error &e)
+      {
+        if (!e.is_out_of_memory())
+          throw;
+        if (++try_count == 2)
+          throw;
+      }
+
+      alloc.try_release_blocks();
+    }
+  }
+
+  // }}}
+
+
+  // {{{ pooled_svm
+
+  class pooled_svm
+    : public pyopencl::pooled_allocation<pyopencl::memory_pool<svm_allocator>>,
+    public pyopencl::svm_pointer
```

Comment:
Should this also become a subclass of `memory_object_holder` (like `pooled_buffer`)? Otherwise, pytato fails with 

```
 File "/shared/home/mdiener/Work/emirge/pytato/pytato/transform/__init__.py", line 1776, in cached_data_wrapper_if_present
    cache_key = _get_data_dedup_cache_key(ary.data)
  File "/shared/home/mdiener/Work/emirge/pytato/pytato/transform/__init__.py", line 1729, in _get_data_dedup_cache_key
    assert (isinstance(ary.base_data, MemoryObjectHolder)
```

Reply:
This should also tolerate `cl.SVMPointer`, though make sure to write the corresponding import in a way that still allows older versions of PyOpenCL that do not have that.